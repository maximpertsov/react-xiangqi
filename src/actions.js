export {
  addMove,
  cancelMoves,
  confirmMoves,
  fetchInitialPlacement,
  fetchMoveInfo,
  makeMove,
  postMove,
  selectMove,
  toggleMovesFetched,
  toggleLoading,
} from 'scenes/Game/actions';

export {
  fetchGames,
  setAutoMoveOff,
  setAutoMoveRed,
  setAutoMoveBlack,
  setAutoMoveBoth,
  setGameSlug,
  setForm,
  setUsername,
  toggleShowGame,
  toggleCanMoveBothColors,
} from 'scenes/Home/actions';

export {
  clearAnimationOffset,
  clearSelectedSlot,
  setAnimationOffset,
  setSelectedSlot,
} from 'components/Board/actions';

export default {};
