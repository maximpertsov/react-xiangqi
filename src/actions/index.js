import { createActions } from 'redux-actions';
import { Color } from 'services/logic/constants';

export default createActions({
  BOARD: {
    ANIMATION_OFFSET: {
      SET: undefined,
      CLEAR: () => [0, 0],
    },
    SELECTED_SQUARE: {
      SET: undefined,
    },
  },
  GAME: {
    CAN_MOVE_BOTH_COLORS: {
      SET: undefined,
    },
    REQUESTED_TAKEBACK: {
      SET: undefined,
    },
    PLAYERS: {
      SET: undefined,
    },
    MOVES: {
      ADD: undefined,
      REMOVE: undefined,
      UPDATE: undefined,
      SET: undefined,
    },
    SELECTED_MOVE: {
      SET: undefined,
    },
    SHOW_CONFIRM_MOVE_MENU: {
      SET: undefined,
    },
    SLUG: {
      SET: undefined,
    },
    UPDATE_COUNT: {
      SET: undefined,
    },
  },
  HOME: {
    AUTO_MOVE: {
      SET: {
        OFF: () => [],
        RED: () => [Color.RED],
        BLACK: () => [Color.BLACK],
        BOTH: () => [Color.RED, Color.BLACK],
      },
    },
    GAMES: {
      SET: undefined,
    },
    LOGIN_FORM: {
      USERNAME: {
        SET: undefined,
      },
      PASSWORD: {
        SET: undefined,
      },
      ERROR: {
        SET: undefined,
      },
      LOADING: {
        SET: undefined,
      },
    },
    SHOW_GAME: {
      SET: undefined,
    },
    USERNAME: {
      SET: undefined,
    },
  },
});
