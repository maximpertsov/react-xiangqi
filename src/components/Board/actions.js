export const setSelectedSlot = ({ selectedSlot }) => ({
  type: 'set_selected_slot',
  selectedSlot,
});

export const clearSelectedSlot = () => setSelectedSlot({ selectedSlot: null });

export const setAnimationOffset = ({
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
}) => ({
  type: 'set_animation_offset',
  bottomPlayerIsRed,
  fromSlot,
  toSlot,
});

export const clearAnimationOffset = () => ({
  type: 'clear_animation_offset',
});

export default {};
