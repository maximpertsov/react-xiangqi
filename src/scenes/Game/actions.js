import update from 'immutability-helper';
import XiangqiBoard, * as logic from 'services/logic';
import { Color } from 'services/logic/constants';

// TODO: find out why this id starts at 3
let nextMoveId = 0;

const newMove = (moveData) => (
  {
    ...moveData,
    id: ++nextMoveId,
  }
);

/*********************/
/*** Initial State ***/
/*********************/

// TODO: define in logic class
/* eslint-disable-next-line max-len */
const DEFAULT_FEN = 'rheakaehr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RHEAKAEHR';

const getInitialMovesState = (fen = DEFAULT_FEN) => ([
  newMove({
    piece: undefined,
    fromPos: undefined,
    toPos: undefined,
    board: new XiangqiBoard({ fen }),
    pending: false,
  }),
]);

const initialPlayers = [
  { name: undefined, color: Color.RED },
  { name: undefined, color: Color.BLACK },
];

export const getInitialState = () => ({
  loading: true,
  moves: getInitialMovesState(),
  moveCount: 0,
  players: initialPlayers,
  selectedMoveId: undefined,
});

/***************/
/*** Actions ***/
/***************/

const incrementMoveCount = (state) => (
  { ...state, moveCount: state.moveCount + 1 }
);

const getMoveIndex = ({ moves }, moveId ) => {
  if (moveId === undefined) return 0;

  return moves.findIndex(({ id }) => id === moveId);
};

const setSelectedMoveIndex = (state, moveIndex) => {
  const { id: selectedMoveId } = state.moves[moveIndex];
  return { ...state, selectedMoveId };
};

export const setSelectedMove = (state, moveId) =>
  setSelectedMoveIndex(state, getMoveIndex(state, moveId));

export const setPreviousMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.max(moveIndex - 1, 0));
};

export const setNextMove = (state) => {
  const moveIndex = getMoveIndex(state, state.selectedMoveId);
  return setSelectedMoveIndex(state, Math.min(moveIndex + 1, state.moveCount));
};

const selectLastMove = (state) =>
  setSelectedMoveIndex(state, state.moves.length - 1);

export const addMove = (state, board, fromSlot, toSlot, pending) => {
  const move = newMove({
    fromPos: logic.getRankFile(fromSlot),
    toPos: logic.getRankFile(toSlot),
    piece: board.getPiece(fromSlot),
    board: board.move(fromSlot, toSlot),
    pending,
  });
  return selectLastMove(
    incrementMoveCount(
      update(state, { moves: { $push: [move] } }),
    ),
  );
};

export const cancelMoves = (state) => {
  const { moves, selectedMoveId } = state;
  const confirmedMoves = moves.filter(({ pending }) => !pending);
  const selectedMoveIndex = Math.min(selectedMoveId, confirmedMoves.length - 1);

  return {
    ...state,
    moves: confirmedMoves,
    moveCount: confirmedMoves.length,
    selectedMoveId: setSelectedMoveIndex(state, selectedMoveIndex),
  };
};

export const confirmMoves = (state) => {
  const { moves } = state;
  return {
    ...state,
    moves: moves.map((move) => ({ ...move, pending: false })),
  };
};

const setMove = (state, { origin: fromPos, destination: toPos }) => {
  const { board } = state.moves[state.moves.length - 1];
  const move = newMove({
    piece: board.getPiece(fromPos, logic.RefType.RANK_FILE),
    fromPos,
    toPos,
    board: board.move(fromPos, toPos, logic.RefType.RANK_FILE),
    pending: false,
  });
  return incrementMoveCount(update(state, { moves: { $push: [move] } }));
};

export const setMoves = (state, moves) =>
  selectLastMove(
    moves.reduce(
      (prevState, move) => setMove(prevState, move),
      {
        ...state,
        moves: getInitialMovesState(),
        moveCount: 0,
        loading: false,
      },
    ),
  );

export default {};
