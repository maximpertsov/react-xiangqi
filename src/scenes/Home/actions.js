import * as client from 'services/client';

export const fetchGames = ({ username }) => async dispatch => {
  if (username === null) return;

  const {
    data: { games },
  } = await client.getGameList({ username });
  dispatch({ type: 'set_games', games });
};

export const setForm = form => dispatch => {
  const { username, password, error } = form;
  if (username !== undefined) {
    dispatch({ type: 'set_login_username', username });
  }
  if (password !== undefined) {
    dispatch({ type: 'set_login_password', password });
  }
  if (error !== undefined) {
    dispatch({ type: 'set_login_error', error });
  }
};

export default {};
