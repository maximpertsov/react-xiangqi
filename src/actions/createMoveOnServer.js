import actions from 'actions';
import client from 'services/client';

const postMove = ({ gameSlug, uci, fen, username }) => {
  // TODO: change `move:` after data model update on server
  const payload = {
    game: gameSlug,
    name: 'move',
    payload: { uci, fen, player: username },
  };
  return client.post(`game/events`, payload);
};

const createMoveOnServer = ({
  gameSlug,
  uci,
  fen,
  username,
}) => async dispatch => {
  if (gameSlug === null) return;

  postMove({ gameSlug, uci, fen, username }).catch(() =>
    dispatch(actions.game.moves.remove(fen)),
  );
};

export default createMoveOnServer;
