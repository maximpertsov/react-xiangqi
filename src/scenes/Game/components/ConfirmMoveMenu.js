import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { cancelMoves, confirmMoves, postMove, selectMove } from 'actions';
import { getLastMove, getPreviousMove } from 'reducers';

import isEqual from 'lodash/isEqual';

import ConfirmMenu from 'components/ConfirmMenu';

const ConfirmMoveMenu = () => {
  const dispatch = useDispatch();

  const gameSlug = useSelector(state => state.gameSlug);
  const lastMove = useSelector(state => getLastMove(state), isEqual);
  const previousMove = useSelector(state => getPreviousMove(state), isEqual);
  const username = useSelector(state => state.username);

  const confirmGameMove = useCallback(() => {
    dispatch(postMove({ gameSlug, move: lastMove, username }));
    dispatch(confirmMoves());
  }, [dispatch, gameSlug, lastMove, username]);

  const confirmMove = useCallback(async () => {
    if (!lastMove.pending) return;

    confirmGameMove();
  }, [confirmGameMove, lastMove.pending]);

  const cancelMove = useCallback(() => {
    // HACK: select previous move before dropping the cancelled move
    dispatch(selectMove({ moveId: previousMove.id }));
    dispatch(cancelMoves());
  }, [dispatch, previousMove.id]);

  return (
    <ConfirmMenu
      label="Confirm move?"
      yesHandler={confirmMove}
      noHandler={cancelMove}
      show={lastMove.pending}
      disabled={gameSlug === null}
    />
  );
};

export default ConfirmMoveMenu;
