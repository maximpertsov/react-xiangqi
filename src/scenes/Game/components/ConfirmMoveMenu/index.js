import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { cancelMoves, confirmMoves, postMove, selectMove } from 'actions';
import { getLastMove, getPreviousMove } from 'reducers';

import ConfirmMenu from 'components/ConfirmMenu';

const ConfirmMoveMenu = () => {
  const dispatch = useDispatch();

  const gameSlug = useSelector(state => state.gameSlug);
  const lastMove = useSelector(state => getLastMove(state));
  const moves = useSelector(state => state.moves);
  const previousMove = useSelector(state => getPreviousMove(state));
  const username = useSelector(state => state.username);

  const confirmGameMove = useCallback(
    move => {
      dispatch(postMove({ gameSlug, move, moves, username }));
      dispatch(confirmMoves());
    },
    [dispatch, gameSlug, moves, username],
  );

  // const confirmFenMove = useCallback(() => {
  //   return;
  // }, []);

  const confirmMove = useCallback(async () => {
    if (!lastMove.pending) return;

    confirmGameMove(lastMove);
  }, [confirmGameMove, lastMove]);

  const cancelMove = useCallback(() => {
    // HACK: select previous move before dropping the cancelled move
    dispatch(selectMove({ moveId: previousMove.id }));
    dispatch(cancelMoves());
  }, [dispatch, previousMove.id]);

  return (
    <ConfirmMenu
      yesHandler={confirmMove}
      noHandler={cancelMove}
      show={lastMove.pending}
      disabled={gameSlug === null}
    />
  );
};

export default ConfirmMoveMenu;
