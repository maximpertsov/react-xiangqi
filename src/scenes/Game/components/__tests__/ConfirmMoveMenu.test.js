import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { mount, render } from 'enzyme';

import { applyMiddleware, createStore, compose } from 'redux';
import rootReducer from 'reducers';
import * as createMoveOnServer from 'actions/createMoveOnServer';

import ConfirmMoveMenu from 'scenes/Game/components/ConfirmMoveMenu';

const initialState = {
  animationOffset: [0, 0],
  gameSlug: 'ABC123',
  showConfirmMoveMenu: true,
  positions: [
    { id: 0, fan: null },
    { id: 1, fan: 'a1a2' },
  ],
  players: [
    {
      color: 'red',
    },
    {
      color: 'black',
    },
  ],
  selectedMoveId: null,
  selectedSquare: null,
  showGame: true,
  username: 'user',
};

const getStore = (overrides = {}) =>
  compose(applyMiddleware(thunk))(createStore)(rootReducer, {
    ...initialState,
    ...overrides,
  });

const getComponent = store => (
  <Provider store={store}>
    <ConfirmMoveMenu />
  </Provider>
);

describe('ConfirmMoveMenu', () => {
  test('renders without crashing', () => {
    const wrapper = render(getComponent(getStore()));
    expect(wrapper).toMatchSnapshot();
  });

  test('confirm closes the method', () => {
    const spy = jest.spyOn(createMoveOnServer, 'default');

    const store = getStore();
    const wrapper = mount(getComponent(store));

    // Click confirm
    wrapper.find('.button.green').simulate('click');
    jest.runOnlyPendingTimers();
    wrapper.update();

    // TODO: Test actions with mockStore

    // Calls API
    expect(spy).toHaveBeenCalledWith({
      gameSlug: 'ABC123',
      id: 1,
      fan: 'a1a2',
      username: 'user',
    });

    // Confirm state
    expect(store.getState().showConfirmMoveMenu).toBe(false);
    expect(wrapper.render()).toMatchSnapshot();

    spy.mockRestore();

    wrapper.unmount();
  });

  test('cancel removes the last move and closes the method', () => {
    const store = getStore();
    const wrapper = mount(getComponent(store));

    // Click confirm
    wrapper.find('.button.red').simulate('click');
    jest.runOnlyPendingTimers();
    wrapper.update();

    // TODO: Test actions with mockStore

    // Confirm state
    expect(store.getState().showConfirmMoveMenu).toBe(false);
    expect(store.getState().positions).toEqual([{ id: 0, fan: null }]);
    expect(wrapper.render()).toMatchSnapshot();

    wrapper.unmount();
  });
});
