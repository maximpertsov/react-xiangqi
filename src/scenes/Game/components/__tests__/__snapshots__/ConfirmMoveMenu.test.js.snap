// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmMoveMenu cancel removes the last move and closes the method 1`] = `null`;

exports[`ConfirmMoveMenu confirm closes the method 1`] = `null`;

exports[`ConfirmMoveMenu renders without crashing 1`] = `
<div
  class="ui segment"
>
  <div
    class="ui top attached label"
  >
    Confirm move?
  </div>
  <div
    class="GameMenu css-1rjz4w9"
  >
    <button
      class="ui green button"
    >
      <i
        aria-hidden="true"
        class="check fitted icon"
      />
    </button>
    <button
      class="ui red button"
    >
      <i
        aria-hidden="true"
        class="x fitted icon"
      />
    </button>
  </div>
</div>
`;
